### DISCUSSION TITLE
Is there a way to install a package that is located at an index-url that uses the keyring authentication method?

### DISCUSSION THREAD

**Discussion by search5** [pypa/pipx#1280] [Q&A] [answered]
You want to install a CLI package from the artifact registry on google cloud, and the google artifact registry relies on the keyrings.google-artifactregistry-auth package for user authentication.

I was able to successfully install the package using keyrings by simply using the pip command or using pipenv, poetry, etc.

However, with pipx, I'm not sure how to install a package that is located at an external index-url that uses the keyring module.

I was wondering if anyone has any idea?

**Comment by dukecat0** [comment]
How about using `--preinstall` feature introduced in https://github.com/pypa/pipx/pull/900?
> Adds support for injecting packages before performing the main package installation. This allows for keyring support via the > following, for example for using the Google Artifact Registry:
>```
>pipx install \
  >--index-url=https://us-python.pkg.dev/my-project/my-index/simple \
  >--preinstall=keyrings.google-artifactregistry-auth \
  >my-private-package
>```

**Reply by search5** [reply]
Thanks for the helpful information, but I'm still failing to install packages from private pypi repositories.

**Reply by dukecat0** [reply]
Could you post the output of `pipx install ... --verbose` here?

**Reply by search5** [reply]
I checked and it looks like the --no-input option is being forced and not installed. It would be nice to have this option by default but be able to exclude it at pipx install.

And on a slightly different note, it would be nice to get extra-index-url instead of index-url on pipx install, because sometimes packages in the private repo might depend on packages in the public pypi...

**Reply by dukecat0** [reply]
> I checked and it looks like the --no-input option is being forced and not installed. It would be nice to have this option by default but be able to exclude it at pipx install.

There are reasons that `--no-input` is passed. See #1029

Could you try https://github.com/pypa/pipx/issues/1200#issuecomment-1887573170 and see if it works for you?

> And on a slightly different note, it would be nice to get extra-index-url instead of index-url on pipx install, because sometimes packages in the private repo might depend on packages in the public pypi...

You can run the following command:
```
pipx install --index-url https://us-python.pkg.dev/my-project/my-index/simple --pip-args='--extra-index-url https://pypi.org/simple/' some-package
```

**Reply by search5** [reply]
First of all, thank you for your kind response.
Unfortunately, when using google artifact registry in my local environment, the --no-input option is enforced, so even after trying the comment in #1200, the package still doesn't install.

I have found that if I deliberately clone pipx and add the --use-input option to temporarily prevent --no-input from being enforced, the package is added normally.

I realize that you can use --pip-args to use the --extra-index-url option, but given that it might be used relatively often, I wonder if it wouldn't be nice to pull it out as a default option in pipx like --index-url.

**Reply by Gitznik** [reply]
`--no-input` disables looking for keychain, that's why removing `--no-input` fixed this for you. I don't think it's the right approach though, instead I would recommend explicitly enabling keychain again via `--keyring-provider`. This fixed it when using pip straight for a package I have on GCP:
`pip install --no-input --keyring-provider import --extra-index-url https://<region>-python.pkg.dev/<registry> <package>`

**Reply by Gitznik** [reply]
@search5 did you have a chance to check this out? Would be great to have a accepted answer for anyone else running into this :)

**Reply by Gitznik** [reply, marked as answer]
I built a small CLI myself and deployed it to GCP.

I was able to install it with pipx via the following command:
```bash
pipx install \
    --index-url=https://<gcp-region>-python.pkg.dev/<project>/<registry-name>/simple/ \
    --preinstall=keyrings.google-artifactregistry-auth \
    --pip-args='--extra-index-url https://pypi.org/simple/ --keyring-provider import' \
    <app-name>
```
Pipx version `1.4.3`.

If you get auth errors, you may have to run one of the following commands first (either should work according to the keyring docs):
```bash
gcloud auth application-default login
gcloud auth login
```

Will mark this as accepted answer and close the discussion. Let me know if you have any other questions.

**Reply by jacksongolden** [reply]
How would you recommend upgrading the package? If I run the analogous `upgrade` command:

```
pipx: error: unrecognized arguments: --preinstall=keyrings.google-artifactregistry-auth
```

**Comment by search5** [comment]
The --preinstall option is available starting with pipx version 1.3.0.
We recommend that you upgrade your version of pipx.
------------------------------------------------------------------------
------------------------------------
There is a future society, but my future is not what others do.
Dept. of Information Science, Graduate School, Korea National Open
University


2024년 7월 31일 (수) 오후 7:42, jake golden ***@***.***>님이 작성:

**Reply by jacksongolden** [reply]
Sorry for lack of clarity. I'm using 1.6, but I'm talking specifially about running `pipx upgrade`.

**Comment by search5** [comment]
What failures are you experiencing when upgrading pipx?
Are you trying to update a package in a private gcp repo?
In any case, try running the pipx upgrade command followed by just the name
of the installed package.

This is because the pipx upgrade command does not have a --preinstall
option.
If you still get the error, please start a new error ticket.
------------------------------------------------------------------------
------------------------------------
There is a future society, but my future is not what others do.
Dept. of Information Science, Graduate School, Korea National Open
University


2024년 8월 1일 (목) 오전 2:17, jake golden ***@***.***>님이 작성:

**Reply by jacksongolden** [reply]
I have installed a package using the keyring authentication method using the solution provided by @Gitznik above.

I have now pushed a new version of _my_ package to Artifact Registry repo and want to upgrade the installed version using `pipx`.

If I run `pipx upgrade my-package`, I get: `my-package is already at latest version x.x.x`, but there is in fact a more recent version on the Artifact Registry repo.

**Comment by search5** [comment]
Try running the “pipx upgrade <package>” command, specifying the new
version uploaded to GCP when you run it.

ex) pipx upgrade sample==0.2
------------------------------------------------------------------------
------------------------------------
There is a future society, but my future is not what others do.
Dept. of Information Science, Graduate School, Korea National Open
University


2024년 8월 1일 (목) 오전 2:27, jake golden ***@***.***>님이 작성:

**Reply by jacksongolden** [reply]
Thanks for the suggestion! Didnt work, sadly:

```
$ pipx ugrade my-package==0.0.2
Package is not installed. Expected to find /Users/.../pipx/venvs/my-package==0.0.2, but it does not exist.
```
